# 1. IPYNB análisis dataset Titanic

In [39]:
import pandas as pd

print(f'Pandas version: {pd.__version__}')

Pandas version: 3.0.0


### 1.1 — Exploración inicial

Muestra las primeras 10 filas del Dataframe

In [40]:
titanic = pd.read_csv('datasets/Titanic-Dataset.csv')

titanic.head(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


Muestra la forma del dataset (filas y columnas) usando ``.shape``.

In [41]:
print('Forma del dataset Titanic-Dataset:', titanic.shape)

Forma del dataset Titanic-Dataset: (891, 12)


Lista los nombres de todas las columnas

In [42]:
# forma 1: titanic.columns

# forma 2:
for i, col in enumerate(titanic.columns, start=1):
    print(f'{i}. {col}')

1. PassengerId
2. Survived
3. Pclass
4. Name
5. Sex
6. Age
7. SibSp
8. Parch
9. Ticket
10. Fare
11. Cabin
12. Embarked


Muestra los tipos de datos de cada columna con ``.dtypes``.

In [43]:
# forma 1:
display(titanic.dtypes)

# forma 2:
'''
for i, (col, dtype) in enumerate(titanic.dtypes.items(), start=1):
    print(f'{i}. {col}: {dtype}')
'''

PassengerId      int64
Survived         int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Cabin              str
Embarked           str
dtype: object

"\nfor i, (col, dtype) in enumerate(titanic.dtypes.items(), start=1):\n    print(f'{i}. {col}: {dtype}')\n"

Usa ``.info()`` para tener una vista general del dataset

In [44]:
titanic.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


### 1.2 — Limpieza de datos

Identifica cuántos valores nulos hay por columna. ¿Qué columnas tienen más valores faltantes?

In [45]:
print(titanic.isnull().sum().sort_values(ascending=False))

Cabin          687
Age            177
Embarked         2
PassengerId      0
Name             0
Pclass           0
Survived         0
Sex              0
Parch            0
SibSp            0
Fare             0
Ticket           0
dtype: int64


Imputa los valores nulos de la columna Age con la mediana de esa columna.

In [46]:
titanic['Age'] = titanic['Age'].fillna(titanic['Age'].median())

Elimina la columna Cabin (tiene demasiados valores nulos para ser útil).

In [47]:
titanic = titanic.drop(columns=['Cabin'])
titanic.head(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S


Elimina las filas que aún tengan valores nulos.

In [48]:
titanic = titanic.dropna()

Verifica que no queden valores nulos usando .isnull().sum().

In [49]:
display(titanic.isnull().sum())

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

### 1.3 — Transformación

Crea una nueva columna llamada FamilySize que sea la columna de SibSp + Parch + 1.

In [50]:
titanic['FamilySize'] = titanic['SibSp'] + titanic['Parch'] + 1

titanic.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,FamilySize
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S,2
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,2
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S,1
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,2
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S,1


Crea una columna IsAlone que valga 1 si FamilySize == 1, y 0 en caso contrario.

In [51]:
titanic['IsAlone'] = titanic['FamilySize'].apply(lambda x: 1 if x == 1 else 0)

titanic.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,FamilySize,IsAlone
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S,2,0
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,2,0
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S,1,1
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,2,0
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S,1,1


Convierte la columna Sex en valores numéricos: male = 0, female = 1.

In [52]:
titanic['Sex'] = titanic['Sex'].apply(lambda x: 0 if x == 'male' else 1)

titanic.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,FamilySize,IsAlone
0,1,0,3,"Braund, Mr. Owen Harris",0,22.0,1,0,A/5 21171,7.2500,S,2,0
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,38.0,1,0,PC 17599,71.2833,C,2,0
2,3,1,3,"Heikkinen, Miss. Laina",1,26.0,0,0,STON/O2. 3101282,7.9250,S,1,1
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,35.0,1,0,113803,53.1000,S,2,0
4,5,0,3,"Allen, Mr. William Henry",0,35.0,0,0,373450,8.0500,S,1,1


### 1.4 — Análisis y preguntas

**P1. ¿Cuál fue la tasa de supervivencia general (porcentaje)?**

In [53]:
tasa_supervivencia = titanic['Survived'].mean() * 100

print(f'Tasa de supervivencia general: {tasa_supervivencia:.2f}%')

Tasa de supervivencia general: 38.25%


*La tasa de supervivencia del 38.25%, para la época, nos indica que el crucero no contaba con los recursos suficientes para salvar más vidas y la arquitectura no era tan buena como parecía*

**P2. ¿Cuál es la tasa de supervivencia según el sexo del pasajero?**

In [54]:
supervivencia_sexo = round(titanic.groupby('Sex')['Survived'].mean() * 100, 2)
print(f'{supervivencia_sexo}\nmale=0, female=1')

Sex
0    18.89
1    74.04
Name: Survived, dtype: float64
male=0, female=1


*Los resultados indican que la tasa de supervivencia de las mujeres fue mayor a la de los hombres. Se peude deber a la prioridad que se le dio a niños y mujeres, además del sacrificio de hombres por sus familias o parejas, como hizo Jack con Rose*

**P3. ¿Qué clase de pasajero (Pclass) tuvo la mayor tasa de supervivencia?**

In [55]:
supervivencia_clase = round(titanic.groupby('Pclass')['Survived'].mean() * 100, 2)
print(f'{supervivencia_clase}\nalta=1, media=2, baja=3')

Pclass
1    62.62
2    47.28
3    24.24
Name: Survived, dtype: float64
alta=1, media=2, baja=3


*Los pasajeros de primera clase son los que mayor tasa de supervivencia obtuvieron, una de las razpones puede ser que estaban mejor ubicados y tenían más información, además eran mejor tratados*

**P4. ¿Cuál es la edad promedio de los sobrevivientes vs. los no sobrevivientes?**

In [56]:
supervivencia_edad = round(titanic.groupby('Survived')['Age'].mean(), 2)
supervivencia_edad.index = ['No sobrevivieron', 'Sobrevivieron']

print(supervivencia_edad)

No sobrevivieron    30.03
Sobrevivieron       28.16
Name: Age, dtype: float64


*Aunque la diferencia no es tan grande, la prioridad fue para los más jóvenes y mujeres*

**P5. ¿Los pasajeros que viajaban solos (IsAlone == 1) tenían mayor o menor probabilidad de sobrevivir?**

In [57]:
supervivencia_individual = round(titanic.groupby('IsAlone')['Survived'].mean() * 100, 2)
supervivencia_individual.index = ['Acompañados', 'Solos']

print(supervivencia_individual)

Acompañados    50.56
Solos          30.09
Name: Survived, dtype: float64


*En este caso si hay más diferencia, los resultados indican que era más probable sobrevivir si se viajaba acompañado, quizá porque no recibieron ayuda del que podría haber sido el/la acompañante*